platform/security/handler/security-handler-pki/src/main/java/org/codice/ddf/security/handler/pki/CrlChecker.java

```diff
@@ -0,0 +1,196 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.security.PropertiesLoader;
+import ddf.security.common.audit.SecurityLogger;
+
+public class CrlChecker {
+
+    public static final String CRL_PROPERTY_KEY = Merlin.OLD_PREFIX + Merlin.X509_CRL_FILE;
+
+    protected static final Logger LOGGER = LoggerFactory.getLogger(AbstractPKIHandler.class);
+
+    private static String encryptionPropertiesLocation = "etc/ws-security/server/encryption.properties";
+
+    private CRL crl = null;
+
+    /**
+     * Constructor method. Reads encryption.properties and sets CRL location
+     */
+    protected CrlChecker() {
+        Properties encryptionProperties = loadProperties(encryptionPropertiesLocation);
+        setCrlLocation(encryptionProperties.getProperty(CRL_PROPERTY_KEY));
+    }
+
+    public HandlerResult check(HttpServletResponse httpResponse, BaseAuthenticationToken token,
+            X509Certificate[] certs, HandlerResult handlerResult) {
+        if (crl == null) {
+            LOGGER.trace("CRL is not set. Skipping CRL check");
+            handlerResult.setToken(token);
+            handlerResult.setStatus(HandlerResult.Status.COMPLETED);
+            return handlerResult;
+        }
+        LOGGER.trace("Checking request certs against CRL.");
+        return checkAgainstCrl(httpResponse, token, certs, handlerResult);
+    }
+
+    /**
+     * Checks the certificates against the CRL. If it is in the CRL, send a 401 error and return a HandlerResult with
+     * the status of REDIRECTED. Otherwise, set appropriate tokens on the HandlerResult and return with status of COMPLETED
+     *
+     * @param httpResponse  HttpServletResponse to send 401 error if needed
+     * @param token         BaseAuthenticationToken containing the auth data to be attached to the HandlerResult if it passes the CRL
+     * @param certs         Certificates extracted from the request to check against the CRL
+     * @param handlerResult HandlerResult to modify and return
+     * @return returns the modified handler result. REDIRECTED status if it failed the CRL check or COMPLETED if it passed
+     */
+
+    HandlerResult checkAgainstCrl(HttpServletResponse httpResponse, BaseAuthenticationToken token,
+            X509Certificate[] certs, HandlerResult handlerResult) {
+        if (passesCRL(certs)) {
+            handlerResult.setToken(token);
+            handlerResult.setStatus(HandlerResult.Status.COMPLETED);
+        } else {
+            // cert is present and in the CRL list - set handlerResult to REDIRECTED and return 401
+            handlerResult.setStatus(HandlerResult.Status.REDIRECTED);
+            try {
+                httpResponse.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+                httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+                httpResponse.flushBuffer();
+            } catch (IOException e) {
+                LOGGER.debug("Unable to send 401 response to client.");
```

Comment:
Shouldn't we throw a servlet exception if we are not able to write a 401 for a user trying to use a revoked certificate?  The web sso filter assumes the handler did everything needed when the status is redirected.